Added --target command line argument #922

MazeChaZer opened the pull request:
There it is, my first contribution to the qutebrowser code base! ;)

The implementation follows what we had discussed in #922.

The-Compiler commented on qutebrowser/app.py:
```diff
@@ -255,7 +256,11 @@ def process_pos_args(args, via_ipc=False, cwd=None):
             win_id = mainwindow.get_window(via_ipc, force_window=True)
         else:
-            win_id = mainwindow.get_window(via_ipc)
+            if via_ipc and target_arg and target_arg != 'auto':
```
With the `and target_arg` you're checking if it's not the default argument, right? If so, please use `and target_arg is not None` instead.

MazeChaZer replied:
There is a little problem:
If the target argument is not given, the running instance recieves an empty string instead of `None`, although `argparse` returns `None` from the argument parsing.
The JSON in the ipc still has `"target_arg": null`, so I think the problem comes from the PyQt signal `got_args`:
```got_args = pyqtSignal(list, str, str)```
When called with `got_args.emit(['example.org'], None, '/temp/dir')` [ipc.py:339], it looks like the slot in the `app` module recieves `(['example.org'], '', '/temp/dir')`
```
server.got_args.connect(lambda args, target_arg, cwd:
                                process_pos_args(args, cwd=cwd, via_ipc=True,
                                                 target_arg=target_arg))
```
[app.py:106]
Now that I check for `None` explicitely, it doesn't work anymore.

The-Compiler replied:
This happens because the slot is specified with a `str` type as argument, which means the Python string will be converted to a `QString`. For historical reasons, [there's a difference between empty and null QStrings](http://doc.qt.io/qt-5/qstring.html#distinction-between-null-and-empty-strings). What happens here is that PyQt converts `None` to a null `QString`, and later converts it back to an empty Python string (because Python strings don't have this distinction, and a null string is also empty).

Can you please add the `target_arg` argument to the docstring here and add a quick explanation of what's happening?

The-Compiler commented on qutebrowser/mainwindow/mainwindow.py:
```diff
@@ -55,7 +55,8 @@ def get_window(via_ipc, force_window=False, force_tab=False):
         return 0
     window_to_raise = None
-    open_target = config.get('general', 'new-instance-open-target')
+    if not open_target:
```
Same here - `if open_target is None:`

The-Compiler commented on qutebrowser/misc/ipc.py:
```diff
@@ -322,6 +323,13 @@ def on_ready_read(self):
+            except KeyError:
+                log.ipc.error("target arg missing: {}".format(decoded.strip()))
+                self._handle_invalid_data()
+                return
```
I think it'd be cleaner to do something like:

```python
for name in ('args', 'target_arg'):
    if name not in json_data:
        log.ipc.error("Missing {}: {}".format(name, decoded.strip()))
        self._handle_invalid_data()
        return
```

and then use `json_data['args']` and `json_data['target_arg']` in the `self.got_args.emit` line. Then you can also get rid of the `pylint: disable=too-many-return-statements` again.

The-Compiler commented on qutebrowser/misc/ipc.py:
```diff
@@ -418,8 +426,8 @@ def _has_legacy_server(name):
-def send_to_running_instance(socketname, command, *, legacy_name=None,
-                             socket=None):
+def send_to_running_instance(socketname, command, target_arg, *,
+                             legacy_name=None, socket=None):
```
Please add `target_arg` after the `*`, i.e. as keyword-only argument.

MazeChaZer replied:
Should it default to `None`?

The-Compiler replied:
Hmm, right... since it's basically a required argument, let's keep it as-is.

The-Compiler commented on qutebrowser/misc/ipc.py:
```diff
@@ -500,6 +509,7 @@ def send_or_listen(args):
     try:
         try:
             sent = send_to_running_instance(socketname, args.command,
+                                            args.target,
```
You'll need to adjust this to `target_arg=args.target,` after the change above.

The-Compiler commented on qutebrowser/misc/ipc.py:
```diff
@@ -513,6 +523,7 @@ def send_or_listen(args):
             log.init.debug("Got AddressInUseError, trying again.")
             time.sleep(0.5)
             sent = send_to_running_instance(socketname, args.command,
+                                            args.target,
```
Same as above.

The-Compiler commented on qutebrowser/qutebrowser.py:
```diff
@@ -66,6 +66,10 @@ def get_argparser():
     parser.add_argument('--json-args', help=argparse.SUPPRESS)
+    parser.add_argument('--target', choices=['auto', 'tab', 'tab-bg',
+                        'tab-silent', 'tab-bg-silent', 'window'],
+                        help="How the urls should be opened if there is "
```
I think "How URLs should be opened ..." would be better here.

MazeChaZer replied:
Yes this sounds better

The-Compiler commented on tests/unit/misc/test_ipc.py:
```diff
@@ -442,8 +442,8 @@ def test_error_immediately(self, ipc_server, caplog):
-        data = '{{"args": ["foo"], "protocol_version": {}}}\n'.format(
-            ipc.PROTOCOL_VERSION)
+        data = '{{"args": ["foo"], "target_arg": "tab", ' \
+               '"protocol_version": {}}}\n'.format(ipc.PROTOCOL_VERSION)
```
Please add parens and drop the `\`, i.e.:

```python
        data = ('{{"args": ["foo"], "target_arg": "tab", '
               '"protocol_version": {}}}\n'.format(ipc.PROTOCOL_VERSION))
```

The-Compiler commented:
Thanks, good job! I commented on some style things.

MazeChaZer commented:
Thanks for the fast feedback.

I fixed all the issues, added the new attributes to the docstrings and did some very minor refactoring in `get_window()` [mainwindow.py]

The-Compiler commented:
And thank you for the fast follow-up!

Looks good, I'll test and merge this either later today or tomorrow morning.

MazeChaZer commented:
Just saw that the ci failed because of two long lines, fixed that
